File: backend/encryption/symmetric/railfence_lib.py
import base64
from .base import SymmetricEncryption

class RailFenceLib(SymmetricEncryption):
    def encrypt(self, plaintext: str, key: bytes) -> str:
        # Anahtarın ilk byte'ını 'ray' sayısı olarak kullanıyoruz (min 2)
        rails = max(2, key[0] % 10) 
        fence = [[] for _ in range(rails)]
        rail = 0
        direction = 1

        for char in plaintext:
            fence[rail].append(char)
            rail += direction
            if rail == rails - 1 or rail == 0:
                direction *= -1
        
        res = "".join(["".join(r) for r in fence])
        return base64.b64encode(res.encode('utf-8')).decode('utf-8')

    def decrypt(self, ciphertext: str, key: bytes) -> str:
        data = base64.b64decode(ciphertext).decode('utf-8')
        rails = max(2, key[0] % 10)
        
        # Zigzag desenini oluştur
        pattern = [None] * len(data)
        fence = [[] for _ in range(rails)]
        rail = 0
        direction = 1
        for i in range(len(data)):
            fence[rail].append(i)
            rail += direction
            if rail == rails - 1 or rail == 0:
                direction *= -1
        
        # Karakterleri desene yerleştir
        idx = 0
        for r in range(rails):
            for pos in fence[r]:
                pattern[pos] = data[idx]
                idx += 1
        
        return "".join(pattern)

    def generate_key(self, key_size: int = None) -> bytes:
        return b'\x03' # 3 raylı bir çit varsayılanı

File: frontend/static/js/encryption/symmetric/railfence_lib.js
class RailFenceLib {
    async encrypt(plaintext, key) {
        const rails = Math.max(2, key[0] % 10 || 3);
        let fence = Array.from({ length: rails }, () => []);
        let rail = 0, direction = 1;

        for (let char of plaintext) {
            fence[rail].push(char);
            rail += direction;
            if (rail === rails - 1 || rail === 0) direction *= -1;
        }

        const res = fence.flat().join("");
        return btoa(unescape(encodeURIComponent(res)));
    }

    async decrypt(ciphertext, key) {
        const data = decodeURIComponent(escape(atob(ciphertext)));
        const rails = Math.max(2, key[0] % 10 || 3);
        const len = data.length;
        let positions = Array.from({ length: rails }, () => []);
        let rail = 0, direction = 1;

        for (let i = 0; i < len; i++) {
            positions[rail].push(i);
            rail += direction;
            if (rail === rails - 1 || rail === 0) direction *= -1;
        }

        let result = new Array(len);
        let charIdx = 0;
        for (let r = 0; r < rails; r++) {
            for (let pos of positions[r]) {
                result[pos] = data[charIdx++];
            }
        }
        return result.join("");
    }

    async generateKey() {
        return new Uint8Array([3]);
    }
}